File: python/complex-numbers/complex_numbers.py
from math import hypot, sin, cos, exp as math_exp


class ComplexNumber(object):
    def __init__(self, real, imaginary):
        self.real = real
        self.imaginary = imaginary

    def __eq__(self, other):
        return self.real == other.real and self.imaginary == other.imaginary

    def __add__(self, other):
        return ComplexNumber(
            self.real + other.real,
            self.imaginary + other.imaginary)

    def __mul__(self, other):
        return ComplexNumber(
            self.real * other.real - self.imaginary * other.imaginary,
            self.imaginary * other.real + self.real * other.imaginary)

    def __sub__(self, other):
        return ComplexNumber(
            self.real - other.real,
            self.imaginary - other.imaginary)

    def __truediv__(self, other):
        denom = other.real ** 2 + other.imaginary ** 2
        return ComplexNumber(
            (self.real * other.real + self.imaginary * other.imaginary) / denom,
            (self.imaginary * other.real - self.real * other.imaginary) / denom)

    def __abs__(self):
        return hypot(self.real, self.imaginary)

    def conjugate(self):
        return ComplexNumber(
            self.real,
            -self.imaginary)

    def exp(self):
        ea = math_exp(self.real)
        return ComplexNumber(
            ea * cos(self.imaginary),
            ea * sin(self.imaginary))

File: csharp/complex-numbers/ComplexNumbers.cs
using System;

public struct ComplexNumber
{
    // These wouldn't have underscores if the public methods below didn't exist.
    private readonly double _Real;
    private readonly double _Imaginary;

    public ComplexNumber(double real, double imaginary)
    {
        _Real = real;
        _Imaginary = imaginary;
    }

    // I probably wouldn't have these public getters, but they're required for the test suite.
    public double Real() => _Real;
    public double Imaginary() => _Imaginary;

    public ComplexNumber Mul(ComplexNumber other) 
        => new ComplexNumber(
            _Real * other._Real - _Imaginary * other._Imaginary,
            _Imaginary * other._Real + _Real * other._Imaginary);

    public ComplexNumber Add(ComplexNumber other) 
        => new ComplexNumber(
            _Real + other._Real,
            _Imaginary + other._Imaginary);

    public ComplexNumber Sub(ComplexNumber other) 
        => new ComplexNumber(
            _Real - other._Real,
            _Imaginary - other._Imaginary);

    public ComplexNumber Div(ComplexNumber other)
    {
        double denominator = other._Real * other._Real + other._Imaginary * other._Imaginary;
        return new ComplexNumber(
            (_Real * other._Real + _Imaginary * other._Imaginary) / denominator,
            (_Imaginary * other._Real - _Real * other._Imaginary) / denominator);
    }

    public double Abs() => Math.Sqrt(_Real * _Real + _Imaginary * _Imaginary);

    public ComplexNumber Conjugate()
        => new ComplexNumber(
            _Real,
            -_Imaginary);

    public ComplexNumber Exp()
    {
        double e_to_a = Math.Exp(_Real);
        return new ComplexNumber(
            e_to_a * Math.Cos(_Imaginary),
            e_to_a * Math.Sin(_Imaginary));
    }
}